docs/RFCS/20191209_optimistic_ddl.md

```diff
@@ -0,0 +1,650 @@
+  
+  - (we could try to explore further cases by fuzzing with generated DDLs and DMLs)
+  
+  - (the effect of violating these cases should be documented)
+
+- Upstream and downstream should have the same sql_mode.
+
+Risks:
+
+- All schemas of every shards should be consistent before and after running a batch of DDL statements. (← please further clarify)
+
+  - or, no DMLs before the schemas could become inconsistent
+  
+  - worst-case fix: re-synchronize from the point where all schemas were still consistent.
+
+Some restrictions are lifted, e.g.
+
+- You can now execute DDLs in different order or combination, as long as the final result converges.
+
+- The upstream MySQL schemas and downstream TiDB schema does not need to exactly match; ensuring upstream ≤ downstream is enough.
+
+## Task management
+
+Programs can go wrong, and we may need to fix the states manually. We will need the following additional commands:
+
+- disable-ddl-propagation
+
+  - Disallows DM master sending DDLs to downstream.
+
+- enable-ddl-propagation
+
+  - Allows DM master sending DDLs to downstream. The schema diff during the period while DDL propagation was disabled are to be executed immediately.
+
+- set-schema --worker 192.168.1.10:8262 --table '`db01`.`tbl02`'
+
+  - Forcibly replaces a table’s schema by a custom “create table” statement. 
+
+Commands involving DDL locks (show-ddl-locks, unlock-ddl-lock, break-ddl-lock) can be removed.
+
+## Scratch pad (please ignore)
+
+---
+
+[VMware Tungsten Replicator](https://pubs.vmware.com/continuent/tungsten-replicator-5.0-oss/index.html) supports replicating MySQL binlog or Oracle CDC to any custom downstream database. The scenario closest to shard merging is [fan-in topology](https://pubs.vmware.com/continuent/tungsten-replicator-5.0-oss/deployment-fanin.html), where multiple masters replicate data to the same slave.
+
+However, in face of online schema change, TR simply recommends [pausing replication to the slave](https://pubs.vmware.com/continuent/tungsten-replicator-5.0-oss/operations-schemachanges.html) when performing the DDLs, which isn’t a useful solution to us.
+
+---
+
+A prior art for this is [Citus](http://docs.citusdata.com/en/v8.3/develop/reference_ddl.html), which only officially supports automatic handling of some special DDL kinds on a distributed table, such as “add column”, “drop index”, etc. Other statements must be [executed manually](http://docs.citusdata.com/en/v8.3/develop/reference_propagation.html#manual-prop) on the shards.
+
+---
+
+We cannot support renaming where a column is both the source and target
+
+```sql
+alter table tbl01 rename column d to e;
+alter table tbl01 rename column c to d;
+```
+
+This would require downstream to accept the DML insert into tbl (d) values (...); with potentially two different types. DM master can prevent such situation by reporting error on the second DDL, by examining the column’s state transition.
```

Comment:
```suggestion
This would require downstream to accept the DML `insert into tbl (d) values (...);` with potentially two different types. DM master can prevent such situation by reporting error on the second DDL, by examining the column’s state transition.
```